shibumi/exclude pattern fix 41

shibumi opened the pull request:
Note: This PR is based on PR #41, I couldn't work on the branch directly, because I am not a project maintainer, so I checkout out PR #41 and opened a new PR instead. With this approach we can still list @ayush159 as author for the code he wrote.

**EDIT: I rewrote everything from scratch**

I am right now working on:

* adding documentation (The Apply... func is missing documentation. The func is being exported, so we should definitely add some docs for it)
* Removing unnecessary things or improving error messages. For example: 

https://github.com/in-toto/in-toto-golang/pull/41/files#diff-73e399e3c84826b077d43cc4cbc98e7eR103


Original PR description is below:

**Fixes issue #33**:

**Description of pull request:**
Used go-git to implement gitignore style pattern parsing. Changed the signature of RecordArtifacts function in runlib.go . Added exclude pattern in method signature. If exclude patterns not available, pass nil.

Please verify and check that the pull request fulfills the following
requirements:

- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

shibumi commented:
@lukpueh If you see anything else I can fix in this code, let me know. We should also have a look on the AppVeyor pipeline.. it seems we are running into problems like in the original PR.

shibumi commented:
@lukpueh oh and I am confused about the `exclude_patterns` variable in the metablock. Is this `exclude_patterns` in the official in-toto specification? If not, we should add it. If yes, then we should probably rename it, because I want to be consistent with the rest and I would like to rename it to `excludePatterns` instead of `exclude_patterns`, because all of our other variables are following the official format. At least in the code, I have reviewed.

lukpueh commented:
💯for WIP/draft PRs. I'll take a closer look when the tests pass. ;) Or if you have any specific questions.

> we can still list @ayush159 as author for the code he wrote

That's very kind of you. :)

> We should also have a look on the AppVeyor pipeline.. it seems we are running into problems like in the original PR.

Could this be related to a merge conflict? Maybe try re-basing on master?

> Is this exclude_patterns in the official in-toto specification

No it's not. It is just a feature of the implementation and I don't think it needs to go into the implementation. Feel free to change the casing so that it feeds go style.

shibumi commented:
> Could this be related to a merge conflict? Maybe try re-basing on master?

I have no idea, when I test this branch locally via `go test ./...` everything works. Github won't show me the link to the appveyor build. `go build -o intoto ./in-toto` works as well.

shibumi commented:
I guess we also want to make sure to use the newest `git-go`.. they changed their remote to:

https://github.com/go-git/go-git

lukpueh commented:
Not sure either but there definitely seems to be a merge conflict. Would you mind trying to re-base this branch on master (resolving the conflicts) and force push. Maybe that makes Appveyor happy.

shibumi commented:
sure!

shibumi commented:
I have rebased successfuly (did a few changes on the go.mod file) and pushed a new build..

The build succeeds now, but I still want to have a look on the rest of the code.

shibumi commented:
I still want to change the gopkg.in remote for git-go to the github one..

shibumi commented:
I've updated the go-git remote + version.. but weird that github.com/go-git/go-git pulls in src-d/go-git/v4 as indirect dependency. But we are definitely using v5 in our code now :)

lukpueh commented:
Let me know when you want me to take a look at this.

shibumi commented on in_toto/runlib_test.go:
```diff
@@ -116,9 +144,10 @@ func TestInTotoRun(t *testing.T) {
+			"materialPaths":   {"demo.layout.template"},
+			"productPaths":    {"foo.tar.gz"},
+			"cmdArgs":         {"sh", "-c", "printf out; printf err >&2"},
+			"excludePatterns": nil,
```
I had a look on this via `go test -cover`, this should be covered.

shibumi commented on in_toto/runlib_test.go:
```diff
@@ -27,12 +27,40 @@ func TestRecordArtifact(t *testing.T) {
+			result, err, expected)
+		// os.RemoveAll("tmpdir")
+	}
+	// os.RemoveAll("tmpdir")
```
I removed the lines, because I think they are not necessary anymore.

shibumi commented:
@lukpueh do you want to have a look on this one? I have successfully rebased on our new master, fixed a few things like formatting and added a new test function :)

shibumi commented:
We should make sure to close the other PR and the issue as well when we merge this PR.

shibumi commented:
Thanks to @ayush159 for his work :)

shibumi commented:
@lukpueh did you have some time to look over this? I would like to get the last two PRs merged, before I start working on the in-toto run implementation. I am afraid of too diverging codes if I just start with in-toto-run now. Would make merging unnecessarily complicated.

lukpueh commented on in_toto/runlib.go:
```diff
@@ -48,6 +49,35 @@ func RecordArtifact(path string) (map[string]interface{}, error) {
+		if err != nil {
+			return nil, err
+		}
+		match := m.Match([]string{path}, fileInfo.IsDir())
```
Why do we care if the path is a dir?

shibumi replied:
This is just how the gitignore matcher works: https://pkg.go.dev/gopkg.in/src-d/go-git.v4/plumbing/format/gitignore?tab=doc#Matcher

With a normal regexp matcher we would not need `fileInfo.IsDir()`: https://pkg.go.dev/gopkg.in/src-d/go-git.v4/plumbing/format/gitignore?tab=doc#Matcher

I guess the former dev chose the gitignore matcher, because it implements most of the filesystem path matching already (different platforms: Linux, Windows, Mac, easier path matching, etc)

lukpueh commented on in_toto/runlib.go:
```diff
@@ -65,9 +95,19 @@ the following format:
+
+func RecordArtifacts(paths []string, excludePatterns []string) (map[string]interface{}, error) {
 	artifacts := make(map[string]interface{})
+	paths, err := ApplyExcludePatterns(paths, excludePatterns)
```
I fear this won't be enough. We also need to apply the patterns on the paths resolved in `filepath.Walk` below.

The `paths` slice might only contain directories that are not matched by a pattern, whereas the files in those directories resolved by `Walk` could match.

You can take a look at the reference implementation, to see how it's done there.

shibumi replied:
Do you have a specific test case in mind which I could add? I will have a look on the python implementation. Would be a shame if we say we have regex support, but actually it doesn't work well :D

shibumi replied:
I think you are right. Wouldn't it make more sense if we would apply the patterns to each string instead? So instead of returning a []string slice we should just check if this path is excluded and if so we should return a true and just move to the next path? This would need some rework of the `ApplyExcludePatterns` func

lukpueh replied:
> Do you have a specific test case in mind which I could add?

One test case I have in mind is a file on disk `"/foo/bar"`, an exclude pattern `"bar"` (which should exclude that file), and a path `"/foo"` in the `paths` slice passed to `RecordArtifacts` and walking over files like `/foo/bar`. 

Btw. I just saw that I described the desired behavior in the commit message of https://github.com/in-toto/in-toto/commit/decf211a492b3da3cf2635410bba4f04559bfd17.

I suggest you also take a look at the tests in the reference implementation (see `tests/test_runlib.py`). and maybe coordinate with @itsbenwu, who is working on a related issue (https://github.com/in-toto/in-toto/issues/355).

> Wouldn't it make more sense if we would apply the patterns to each string instead?

Maybe. I think when when I first wrote this in https://github.com/in-toto/in-toto/commit/3ee8b67ea813830c4e228e82b6d85a10174c0752 I saw that the function I used, i.e. `fnmatch.filter` took multiple names as first argument and multiple names was what I had. :) Performance-wise it's many lightweight calls to the filter function vs. one probably more expensive call. Not sure which one performs better. Feel free to profile. To me either approach seems fine.

shibumi commented on in_toto/runlib.go:
```diff
@@ -48,6 +49,35 @@ func RecordArtifact(path string) (map[string]interface{}, error) {
+	}
+	m := NewMatcher(ps)
+	for _, path := range paths {
+		fileInfo, err := os.Stat(path)
```
No, because we are dealing with a gitignore matcher here. This matcher needs to know if we are dealing with a directory or not: https://pkg.go.dev/gopkg.in/src-d/go-git.v4/plumbing/format/gitignore?tab=doc#Matcher

That's why we need FileInfo to retrieve information about the actual state of the file (directory or not).

lukpueh replied:
Thanks for the clarification! This also answers my question below. Would you mind adding a comment?

shibumi replied:
yep

shibumi commented:
> May I suggest you take a look at Table Driven Tests? They seem common in go and we already do this in a bunch of places in this repo. It is a great way to remove redundant test code and to easily add new scenarios.

Sorry, just read this comment now. Table driven tests looks good to me, I think you are talking about something like this:

```
package main

import "testing"

func TestFaculty(t *testing.T) {
	tables := map[uint]uint{
		6: 720,
		2: 2,
		1: 1,
		9: 362880,
	}
	for k, v := range tables {
		result := faculty(k)
		if v != result {
			t.Errorf("Faculty %d! was incorrect, got: %d, want: %d", k, result, v)
		}
	}

}```

Right?

lukpueh commented:
I just thought that [runlib_test.go#L71-L137](https://github.com/in-toto/in-toto-golang/blob/b08a29f19ac2d80dd049c5a45045dd9b0d588e00/in_toto/runlib_test.go#L71-L137)  could easily be written in table form, since each test scenario calls `RecordArtifacts` and then `DeepEqual` and only the parameters differ. But yeah, re-writing the tests is not a big priority.

Btw. I recommend watching [this talk about testing in go](https://youtu.be/8hQG7QlcLBk?t=455). I learned a thing or two from it. :)

shibumi commented:
@lukpueh ah ok, this is what you mean. This shouldn't be much work, I will refactor this section on the weekend. Should be just 10min. The talk looks great! Thanks, I will give it a shot, too.

shibumi commented:
@lukpueh I've refactored the tests for the pattern matching in test tables. The code is much more readable now and it should take us less hassle in the future to add another test if we need to.

shibumi commented:
The latest test case shows indeed that a path like `/$test_dir/foo.tar.gz` will not get excluded if we provide only `foo.tar.gz` as exclude pattern

shibumi commented:
I have reported the gitignore behavior in the go-git repository: https://github.com/go-git/go-git/issues/108

and I am really disappointed, because there is really no magic behind this module and it's actually just calling filepath.Match:

https://golang.org/pkg/path/filepath/#Match

There seems to be no logic in the module for matching on paths via a simple filename pattern: https://github.com/go-git/go-git/blob/master/plumbing/format/gitignore/pattern.go#L92

Also even a pattern like `**/foobar.tar.gz` doesn't seem to work, although it should. I am not sure how we want to proceed in this case. We have multiple options:

1. We could apply `filepath.Base` to the path and then apply the patterns on the last element: https://golang.org/pkg/path/filepath/#Base  Problem: This would not satisfy patterns like `src/pattern.go` in paths like `/tmp/src/pattern.go`.

2. We could try to fix the behavior upstream in go-git. This would freeze this PR for now, because we need to wait for the devs.

3. We could write our own regex module

4. We could try replacing the go-git module with one of these: https://github.com/monochromegane/go-gitignore , https://github.com/denormal/go-gitignore (the latter seems very old..)

lukpueh commented:
Thanks for reporting this upstream, @shibumi and for providing different viable ways to solve this. I lean towards 2 or 4 (with the more active project). 1. Seems a bit hackish and 3. overkill. Note that we don't have to use .gitignore-style matching. The exclude pattern option is not a covered by the spec, it's just a convenient implementation feature.

shibumi commented:
@lukpueh I had deeper look on 4 yesterday and the more active go-gitignore project states in their readme, that they don't support the `**` operator (and this should mean that our path matching isn't supported, too).

But I am confused that the go-git gitignore doesn't support `**`. They state, they do.

lukpueh commented:
@shibumi, would you kindly remove the [corresponding todo item in README](https://github.com/in-toto/in-toto-golang/blame/master/README.md#L40) with this PR?

shibumi commented:
@lukpueh sure, will have a look on it.

shibumi commented:
@lukpueh done, btw

shibumi commented:
@lukpueh I reset this PR to current master and placed my new draft for excludePatterns on top of it. The new pathspec library should solve most of our problems, that got addressed in the comments here. What we need to do now is providing test cases and test the new functionality as good as possible.

shibumi commented:
@lukpueh We should have a look at this PR next. I would like to wait with the Go 1.15 PR, because other go libs (tekton, kubesec etc are maybe not yet ready for Go 1.15)

shibumi commented:
ahh great .. the joy of Windows as only operating system using the backslash as path separator. I will fix this in the pathspec library, then this should be fixed here, as well.

@lukpueh feel free to take a look already. https://github.com/in-toto/in-toto-golang/pull/53/commits/b1cda26262e8084e5a9f5ef48d976f0356a8e057 removes my old approach via ApplyGitIgnore function, because we need to check each path we are visiting via filewalk (I think you mentioned this once, but I totally forgot about this).

The new approach just calls `pathspec.GitIgnore` now directly. This works already in Linux and I test for full paths, "**" operator and "filename" match right now.

I think I will just fix the path separator in the pathspec function via checking if we are on Windows and then translating all "\" to "/", but I don't know yet if this will work.

shibumi commented:
@lukpueh  as promised.. I fixed the issue in upstream go-pathspec.. it works on windows now, too :)

SantiagoTorres commented:
LGTM!

shibumi commented on go.mod:
```diff
@@ -1,5 +1,8 @@
-module github.com/in-toto/in-toto-golang
+module github.com/shibumi/in-toto-golang
```
@SantiagoTorres  oops.. we might want to fix this. No idea how that change got in my commit :D 
I will compile a PR for this tomorrow (too late in Germany already).

Same for the go version
